Processing Module v1.0

sakshi1603 opened the pull request:
This PR includes complete implementation of Processing Module including server side processing (main and board server), processor class, threading, shapes, handlers, logs, exception handling. 
Closes #106 
Closes #100
Closes #95
Closes #93 
Closes #86 
Closes #53
Closes #118

sravand123 committed:
added Team Lead Spec

sravand123 committed:
changed Folder name

sravand123 committed:
Merge pull request #1 from sahul343/SendQueueListener-111701028_Sirpa_Sahul

Added SendQueueListener Specs

sunny117 committed:
add specs for team member - 111701017

shiva-6 committed:
Add Spec for SocketListener

sravand123 committed:
Merge pull request #2 from sunny117/#25-Networking-111701017-Spec

Team Member Spec - 111701017

sravand123 committed:
Merge pull request #7 from shiva-6/#29-SocketListener

Add Spec for SocketListener

sravand123 committed:
Merge pull request #6 from Drisya-Ponmari/#27-receivequeue

ReceiveQueueListner_111701031

sravand123 committed:
Merge pull request #3 from shiva-santhosh/#26-QueueManagement-Design-Spec

Queue-Management-Design-Spec

sravand123 committed:
Merge pull request #9 from shiva-santhosh/#39-Queue-Interface

Added Queue Interface

sravand123 committed:
Merge pull request #8 from Prudhvis-gitrepo/31-111701023-spec

Add Team-member spec & interface-111701023

sravand123 committed:
Merge pull request #10 from Prudhvis-gitrepo/31-111701023-spec

Added INotificationHandler interface

sunny117 committed:
skeleton of LanCommunicator

sunny117 committed:
add subscibeForNotifications func skeleton in class

sravand123 committed:
Merge pull request #13 from sunny117/#25-Networking-111701017-Spec

#44 Skeleton of LanCommunicator Class

sravand123 committed:
Merge pull request #16 from Prudhvis-gitrepo/47-111701023-communicatorfactory

Skeleton of CommunicatorFactory class

sunny117 committed:
Update LanCommunicator.java

sravand123 committed:
Merge pull request #20 from sunny117/#25-Networking-111701017-Spec

Update LanCommunicator.java

sravand123 committed:
Merge pull request #26 from Prudhvis-gitrepo/63-111701023-newfeature1

add new feature to CommunicatoryFactory class

Commit:
Merge branch 'initialise_dirs' of https://github.com/sravand123/KoraKaagaz into basic-server

Commit:
server state added

Commit:
Merge branch 'serialization-apis' of https://github.com/tymefighter/KoraKaagaz into basic-server

Commit:
add client handlers

Commit:
start in CBS

Commit:
Complete existing board request

Commit:
complete new board request

Commit:
Board Server complete

Commit:
while loop in start

Commit:
Remove networking pull

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Remove conflicts

Commit:
Some Changes

Commit:
delete IUser

Commit:
Remove conflicts

Commit:
Add author in handlers

Commit:
Solve conflicts

Commit:
remove handlers

Commit:
Resolve conflicts

tymefighter committed:
Add Visualizer Tool

tymefighter committed:
Add doc

tymefighter committed:
Add doc

tymefighter committed:
Remove Application Exiting

tymefighter committed:
Remove import

tymefighter committed:
Add disposal of object

tymefighter committed:
Update constant

tymefighter committed:
update

tymefighter committed:
update

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
delete to solve conflicts

Commit:
pull

Commit:
solve

Commit:
remove change

Commit:
Remove errors

Commit:
Solve errors

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Remove changed file

Commit:
small change

Commit:
Add comments in CBS

Commit:
Add comments and threading

Commit:
pull

Commit:
add comments in board server

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Merge branch 'week1' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
Add comments

Commit:
Add all the comments

Commit:
Add logs

sakshi1603 committed:
Merge pull request #62 from tymefighter/visualizer-tool

Add Visualizer Tool

tymefighter committed:
Add hashCode methods (#75)

Add hashCode in Utility classes

tymefighter committed:
Shape Drawing Class and Functions (#31)

* Add Shape Builder Class

* Add file structure

* Add rectangle drawer

* Add Shape Helper and Triangle

* Add DDA algorithm

* Add Bresenhams algo

* Update doc

* Update doc

* Update doc

* Add helper API

* Update helpers

* Update helper methods

* Add devansh circle algorithm

* Add doc for rectangle drawer

* Begin adding doc for shape helpers

* Update shape helper doc

* Update shape helper doc

* Add doc for Triangle

* Add methods and doc

* Add doc for circle method

* Add Builder methods defintions

* Add mid point circle drawing algorithm

* Add mid point based circle filling algorithm

* Replace if-else with switch-case

* Minor updates to doc

* Update Line Drawer to use switch case

* Add doc

* Update qualifier, required for visualize

* Fix bug

* Correct comment spelling

* Fix minor bug

tymefighter committed:
Persistance support (#51)

* Add persistance support

* Add encoding

* Add doc

* Update signature and doc

* Update doc

* Update reviwer names

* Update signature

Commit:
Merge branch 'week2' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
add persistance support

tymefighter committed:
Visualize shapes (#64)

Testing the Line, Circle, Triangle Drawing Algorithms by visualizing their outputs
* Add Shape Builder Class
* Add file structure
* Add rectangle drawer
* Add Shape Helper and Triangle
* Add DDA algorithm
* Add Bresenhams algo
* Update doc
* Update doc
* Update doc
* Add helper API
* Update helpers
* Update helper methods
* Add devansh circle algorithm
* Add doc for rectangle drawer
* Begin adding doc for shape helpers
* Update shape helper doc
* Update shape helper doc
* Add doc for Triangle
* Add methods and doc
* Add doc for circle method
* Add Builder methods defintions
* Add mid point circle drawing algorithm
* Add mid point based circle filling algorithm
* Replace if-else with switch-case
* Minor updates to doc
* Update Line Drawer to use switch case
* Add doc
* Add Visualizer Tool
* Add doc
* Add doc
* .
* Update qualifier, required for visualize
* Add Visualizer for Shapes
* Remove Application Exiting
* Remove import
* Visualize Circle
* Minor bugs in line drawing detected
* Add disposal of object
* Add visualizer for shapes
* Fix bug
* Correct comment spelling
* Update constant
* update
* update
* Fix minor bug
* Add author and reviewer
* Minor fix
* Minor fix

Commit:
spelling change

Commit:
Solve Remove Board bug

Commit:
port change

Commit:
small bug

himanshu2047 committed:
Server Side Processing (#54)

Contains complete processing related main server and board server.

Commit:
Merge branch 'week2' of https://github.com/sakshi1603/KoraKaagaz into basic-server

Commit:
bug fix

sakshi1603 committed:
Merge pull request #81 from himanshu2047/basic-server

Basic server

Commit:
Resolve conflict

satchit24 committed:
Screen Clear - Reset Class  (#80)

* Minor Changes

* Add Reset class

* Class Reset Documentation

* Revise Reset

* Reset Class

* Exception handling and Logger

iamrakesh28 committed:
Add logs and bug fix in Undo Redo 

* Add log statements

* Correct indentation in github editor

* Attempt to match eclipse and github indentation

* Test Fix a indentation

* Small indentation fix

* Fix indentation in local editor

* Shift switch case statements to next line

* Update UndoRedo.java

* Update UndoRedo.java

* Fix indentation and spaces

* Add return object

* Fix indentation

* Clear redo stack on changes to maintain consistency

* Fix a bug

* Add log statement for empty stack operation

* Add thread id in log messages

* Fix spacings between functions

* Fix minor indentation

RathoreDevansh08 committed:
Logger, Exception Handling, Design Change (#87)

Add logs, exception handling and return old object in operation util

iamrakesh28 committed:
User API (Add Logger + Fix spacing) (#43)

Add function  User API, Add logs, exception handling and fix indentation

satchit24 committed:
Completed all handlers, Add logs, exception handling (#89)

* Minor Changes

* Add Reset class

* Class Reset Documentation

* Revise Reset

* Reset Class

* Exception handling and Logger

* Handlers with exception and logger

* Handlers completed

* Comments in Object Handler

* Made Corrections

* Intendation changes

* Intendation changes

* Intendation changes

wildcat26 committed:
Wrapper classes for Client Side Threading (#82)

* Create skeleton of wrapper classes for threading
* Implement Select, Delete wrapper classes
* Fix use of ServerCommunicator, add reviewer
* IOperation functions' wrapper classes except for Reset
* Wrapper for Undo Redo operations
* Add TODO in undo-redo, begin DrawCurve
* Call helper classes and send the object to the board server in my thread
* Not sending null object to board server, add javadocs
* Add Exception Handling, Logging to IDrawShapes
* Add helper class to handle exceptions that may arise in logging and sending, use the helper class in other class' run methods
* Add wrapper classes for threading IUser interface methods
* Use Reset API for Reset wrapper in threading
* Fix Reset: send reset object to board server

himanshu2047 committed:
Complete implementation of processor class (#88)

* Create skeleton of wrapper classes for threading

* Create skeleton of wrapper classes for threading

* Implement Select, Delete wrapper classes

* Fix use of ServerCommunicator, add reviewer

* IOperation functions' wrapper classes except for Reset

* Wrapper for Undo Redo operations

* Add TODO in undo-redo, begin DrawCurve

* Call helper classes and send the object to the board server in my thread

* change username to userId

* Add exception handling

* deleted old threading files

* add some code in processor

* processor almost done

* add function for passing boardState to UI

* add thread ID in logs

* add brush radius and board dimensions

* Corrected thead ID in logs

* Deleted Shruti's files

* username to string add

* add return in user functions

* return from reset

* small fix

tymefighter committed:
Add Log Statements (#78)

* Add Log Statements

* Add log for Board Object

* Update log statements

* Add thread ID

* Add reviewer name

* Update thread i

* Update thread id

* Update comments

tymefighter committed:
Algorithms for shapes (#90)

* Add Shape Builder Class

* Add file structure

* Add rectangle drawer

* Add Shape Helper and Triangle

* Add DDA algorithm

* Add Bresenhams algo

* Add helper API

* Update helpers

* Update helper methods

* Add devansh circle algorithm

* Add doc for rectangle drawer

* Begin adding doc for shape helpers

* Update shape helper doc

* Update shape helper doc

* Add doc for Triangle

* Add methods and doc

* Add doc for circle method

* Add Builder methods defintions

* Add mid point circle drawing algorithm

* Add mid point based circle filling algorithm

wildcat26 committed:
Fix Delete to return old object with operation and userId [Exceptions, Logger] (#84)

* Fix Delete to return old object with correct operation and userId
* Add try catch blocks and logging
* Add Exception handling and logs
* Add Exception handling and logging in SelectDelete functions

himanshu2047 committed:
Small Fixes

Fixed indentation errors in files of Ahmed

himanshu2047 committed:
Send the server port to content module

Commit:
Merge with master

muhammedyaseen-git commented:
fine with me, awaiting your thoughts @convict-git ...

convict-git commented:
> fine with me, awaiting your thoughts @convict-git ...

There's something wrong with the PR. Commits from the networking team coming up. ~~Also, 7.5K additions which don't make sense as it just should have mostly loggers and exception handling added after v0.1 (which is merged already)~~. Sorry, for this reply. 

Not sure how is it fine with you! :sweat_smile:

sakshi1603 commented:
> > fine with me, awaiting your thoughts @convict-git ...
> 
> There's something wrong with the PR. Commits from the networking team coming up. Also, 7.5K additions which don't make sense as it just should have mostly loggers and exception handling added after v0.1 (which is merge already).
> 
> Not sure how is it fine with you!

It does not only contain loggers and exception handling. As I mentioned above in PR description, this PR contains: Full Server Side Processing, Processor Class, Threading, Shapes, Handlers, visualiser (Test for Shapes made by @tymefighter for shapes) including logs and try catch.
Networking commits are here because there was delay in merging networking module interfaces. As most of @himanshu2047 work had dependency on networking module, he had pulled the interfaces directly from their PR but when the interfaces were merged he deleted those files and pulled from my master.

tymefighter committed:
Fix Serializability Errors (#95)

Make utility classes serialisable and fix a bug in boardObject

wildcat26 committed:
Change select to return array of Pixels instead of array of Positions (#96)

* Make select return arraylist of Pixels instead of Positions

* Use array of Pixels in threading of select operation

himanshu2047 committed:
Add drawSquare and return type changed in select API

* Implement drawSqaure API in Processor Class.
* Return ArrayList<Pixel> in select API instead of ArrayList<Position>

Commit:
Merge branch 'week3' of https://github.com/sakshi1603/KoraKaagaz into week3

himanshu2047 committed:
Fix while loop bug (#103)

* set userId in ClientBoardState
* Add Thread.sleep(50) in while loop
* Get IP of server in giveUserDetails

wildcat26 committed:
Check if selectedObject is null before using its id (#104)

Co-authored-by: Shruti Umat <111701027@smail.iitpkd.ac.in>

tymefighter committed:
Update timestamp toString method 

return time till milli seconds precision as there is a possibility of two objects having same timestamp

wildcat26 committed:
Override equals() for PriorityQueueObject to fix removing of elements from priority queue 

Override equals() for PriorityQueueObject, used in priority queue while removing object from priority queue (delete function).

wildcat26 committed:
Use Math.round instead of typecasting float to int in DrawSquare 

Improve precision

tymefighter committed:
Triangle line swap fix (#108)

* Update timestamp toString method
* Fix Line Drawer

convict-git commented on src/processing/utility/PriorityQueueObject.java:
```diff
@@ -24,4 +29,23 @@ public PriorityQueueObject(PriorityQueueObject obj) {
+				objectId.equals(p.objectId);
+		}
+		else
+			return false;
```
Missing `}` for the method `equals`. I can resolve this issue at my end, but then it might create conflict issues as we are expecting more changes to come (in order to resolve a few bugs that you mentioned). 
So it's better the authors get it fixed.

sakshi1603 replied:
Somehow you checked outdated file. I had fixed it in the last commit.

convict-git replied:
Actually, that was a comment before the fix, but I forgot to submit the review request. So it shows 9 hours ago but it's from two days ago. Thanks though!

Commit:
minor fix

convict-git commented on src/processing/server/board/ServerCommunication.java:
```diff
@@ -0,0 +1,149 @@
+		 * 
+		 * We are sending username and the client full address
+		 */
+		String message = ClientBoardState.userId.toString()
+					   + ":"
+					   + ClientBoardState.userIP.toString()
+					   + ":"
+					   + ClientBoardState.userPort.toString();
```
Can we use JSON instead? It will give us more flexibility later on, to add more fields if required.

himanshu2047 replied:
Okay

convict-git commented on src/processing/server/main/Requests.java:
```diff
@@ -0,0 +1,112 @@
+		 * While requesting for existing board we need to give boardID and the client's full
+		 * address as the message thus calculating that and combining with delimiter ":"
+		 */
+		String message = boardId.toString()
+					   + ":"
+					   + ipAddress.toString()
+					   + ":"
+					   + portNumber.toString();
```
JSON is preferred here as well.

convict-git commented on src/processing/shape/LineDrawer.java:
```diff
@@ -0,0 +1,404 @@
+	 * 
+	 * Algorithm Link:
+	 * 
+	 * @see <a href="https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm</a>
```
Seems like you missed to close the `a` tag and quotes.

sakshi1603 replied:
Fixed.

convict-git commented on src/processing/server/board/BoardStateRequestHandler.java:
```diff
@@ -0,0 +1,93 @@
+				+ "BoardState request is made by the client"
+		);
+		
+		String[] arrOfArguments = message.split(":", 2);
+		
+		// extract userID from the first part of the message
+		String user = arrOfArguments[1];
+		String[] id = user.split("_", 2);
+		UserId userId = new UserId(new IpAddress(id[0]), new Username(id[1]));
+		
+		/**
+		 * While requesting for BoardState the client passes it's address
+		 * with port as the argument.
+		 */
+		IpAddress clientAddress = new IpAddress(arrOfArguments[1]);
+		
```
All this manipulation will become super annoying even if there's a requirement to add one extra field. Hence choosing JSON as a protocol will be much meaningful choice.

Commit:
Close a tag in comment

himanshu2047 committed:
Add JSON in server communication 

* add json in communication

wildcat26 committed:
Access maps directly in SelectDetete

Remove static variable for maps and access map directly from ClientBoardState

Rahul-R5 commented:
@sakshi1603 @himanshu2047 Does this PR concern server initialization? For server part of content module, ContentFactory can be imported and getContentServer() can be called(static method) to get the server side handler. This is subscribed to the networking module(server side) with identifier "content"

himanshu2047 commented:
> @sakshi1603 @himanshu2047 Does this PR concern server initialization? For server part of content module, ContentFactory can be imported and getContentServer() can be called(static method) to get the server side handler. This is subscribed to the networking module(server side) with identifier "content"

Yes both the server initialisation is there and you can provide me the startContentServer function in your factory class as discussed.

iamrakesh28 committed:
Initialize CBS data members

himanshu2047 committed:
Remove direct initialisation 

* Remove initialisation directly
* call content server
* port bug fix

himanshu2047 commented:
Can be merged now, but suggest to merge it after merging content module otherwise will give build errors. @convict-git @muhammedyaseen-git

himanshu2047 committed:
Add try catch in initialising content server

* Add exception handling while starting content server

* remove warning in Parameterized util
